ttnn/cpp/ttnn/operations/conv/conv2d/prepare_conv2d_weights.cpp

```diff
@@ -0,0 +1,366 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "conv2d.hpp"
+#include <sys/types.h>
+#include <cstdint>
+
+using namespace tt;
+namespace ttnn {
+namespace operations::conv {
+using sliding_window::SlidingWindowConfig;
+using sliding_window::ParallelConfig;
+
+namespace conv2d {
+
+void validate_weight_tensor(const ttnn::Tensor& weight_tensor) {
+    TT_ASSERT(!ttnn::has_storage_type_of(weight_tensor, ttnn::DEVICE_STORAGE_TYPE));
+    TT_ASSERT(weight_tensor.get_layout() == Layout::ROW_MAJOR);
+    TT_ASSERT(weight_tensor.get_shape().rank() == 4);
+}
+
+void validate_bias_tensor(const ttnn::Tensor& bias_tensor) {
+    TT_ASSERT(!ttnn::has_storage_type_of(bias_tensor, ttnn::DEVICE_STORAGE_TYPE));
+    TT_ASSERT(bias_tensor.get_shape().rank() == 4);
+    TT_ASSERT(bias_tensor.get_layout() == Layout::ROW_MAJOR);
+}
+
+void validate_weights_format(std::string weights_format) {
+    TT_FATAL(weights_format.size() == 4, "weights_format must have exactly 4 characters");
+    TT_ASSERT(weights_format.find("O") != string::npos, "weights_format must contain \"O\"");
+    TT_ASSERT(weights_format.find("I") != string::npos, "weights_format must contain \"I\"");
+    TT_ASSERT(weights_format.find("H") != string::npos, "weights_format must contain \"H\"");
+    TT_ASSERT(weights_format.find("W") != string::npos, "weights_format must contain \"W\"");
+    TT_ASSERT(weights_format == "OIHW", "Conv2d weights format must be \"OIHW\"");
+}
+
+template <typename T>
+OptimizedConvBlockConfig get_opt_block_config(
+    bool mm_conv,
+    uint32_t in_channels,
+    uint32_t out_channels,
+    uint32_t output_height,
+    uint32_t output_width,
+    uint32_t batch_size,
+    uint32_t input_width,
+    std::array<uint32_t, 2> kernel_size,
+    std::array<uint32_t, 2> stride,
+    T *device,
+    Conv2dConfig& conv_config) {
+
+    adjust_conv_op_config_for_auto_shard(
```

Comment:
why is this bit always called, it was designed to be called only if input tensor is not sharded and sharding layout is not specified in conv2dconfig

Reply:
I think this happens because again we have c/p blocks of code in there different methods.

Reply:
Well we don't pass the input tensor to the weight preparation function(s). Unless you want me to but I think it would be strange.

Reply:
It's just going to do the wrong thing, you need to pass some info about the input tensor, 
since op will make decisions based on that.

Reply:
try to match what is happening on "main" codepath